### ISSUE TITLE
focal_loss API

### ISSUE THREAD

**Issue by heiyuxiaokai** [TuSimple/simpledet#98]
When I tried to test the focal loss API, it confused me.
cls_label is (1, 22300), cls_label_onehot is (1, 22300, 15)
The test code is: 
![a7bd39ffb38890cbd45d366f92636a7](https://user-images.githubusercontent.com/27018045/58424989-45d7d580-80cb-11e9-90e2-8a0ab4fac0c3.jpg)
The focal loss result:
![dc9f9188c96e6da1cb6bf340ccc2e5d](https://user-images.githubusercontent.com/27018045/58425040-66079480-80cb-11e9-80f0-f0fa9534746f.jpg)
The cls_label_onehot:
![08f974a81bf924f0e06e829ee146857](https://user-images.githubusercontent.com/27018045/58425088-86375380-80cb-11e9-879c-67922322171c.jpg)

It seems more like sigmoid function instead of focal loss fuction? Or the way I test was wrong?

**Comment by heiyuxiaokai** [comment]
When I trained the retinaNet, the FGACC was 0.97 at the epoch of 8.
But at the beginning, the result is about 0.01, it seems like sigmoid(-4.59). -4.59 is the origin cls_bias_init. 
It trains well. But where is the focal loss function?

**Comment by xchani** [comment]
You can find that the **FocalLoss** operator acts as **Sigmoid** in **forward** (due to the laziness of developer :) ) and assigns gradient of **FocalLoss** in **backward**, if you take a good look at [*operator_cxx/contrib/focal_loss-inl.h*](https://github.com/TuSimple/simpledet/blob/master/operator_cxx/contrib/focal_loss-inl.h).

**Comment by heiyuxiaokai** [comment]
Thanks. @xchani 
So the focal loss fuction returns results of forward. I should caculate and print focal loss in metric.py if I want to print focal loss. And other losses are the same.

**Comment by xchani** [comment]
Yes. For other losses, it depends on their implementation details.

**Comment by heiyuxiaokai** [comment]
A question about mxnet.
#cls_logit_concat （N, H*W, C）, label_onehot （N, H*W, C）
pred_logic = mx.sym.sigmoid(cls_logit_concat)
focal_loss = -((1-label_onehot) * (1-p.focal_loss.alpha) * mx.sym.pow(pred_logic, p.focal_loss.gamma) * 
                                  mx.sym.log(1-pred_logic+0.001) + \
                   label_onehot * p.focal_loss.alpha * mx.sym.pow(1 - pred_logic, p.focal_loss.gamma) * 
                                  mx.sym.log(pred_logic+0.001))
total_loss = mx.sym.makeloss(focal_loss)
Does the focal loss of using makeloss function work as well as the function that you define in focal_loss-in.h?

**Comment by xchani** [comment]
Seems good to me. You can check the corretness by feeding the same input and comparing the output and gradient with **FocalLoss** operator. It is worth noticing the **numerical stability** of gradient for this sort of combined symbol.

**Comment by heiyuxiaokai** [comment]
Thanks. I'll test these two operator.
